File: server/routers/analytics.py
from datetime import datetime

import aiosqlite
from fastapi import APIRouter, Depends, Query

from db import get_db
from models.analytics import DailyActiveUsersPoint, HourlyActiveUsersPoint, JoinViolationRankOut, PlayerRankOut
from repositories import analytics as repo

router = APIRouter(prefix="/api/analytics", tags=["analytics"])


@router.get("/daily-active-users", response_model=list[DailyActiveUsersPoint])
async def get_daily_active_users(
    world_id: str | None = Query(None),
    group_id: str | None = Query(None),
    start: datetime | None = Query(None),
    end: datetime | None = Query(None),
    db: aiosqlite.Connection = Depends(get_db),
) -> list[DailyActiveUsersPoint]:
    return await repo.get_daily_active_users(db, world_id, group_id, start, end)


@router.get("/hourly-active-users", response_model=list[HourlyActiveUsersPoint])
async def get_hourly_active_users(
    world_id: str | None = Query(None),
    group_id: str | None = Query(None),
    start: datetime | None = Query(None),
    end: datetime | None = Query(None),
    db: aiosqlite.Connection = Depends(get_db),
) -> list[HourlyActiveUsersPoint]:
    return await repo.get_hourly_active_users(db, world_id, group_id, start, end)


@router.get("/join-violation-rankings", response_model=list[JoinViolationRankOut])
async def get_join_violation_rankings(
    group_id: str = Query(...),
    start: datetime | None = Query(None),
    end: datetime | None = Query(None),
    order: str = Query(default="desc", pattern="^(asc|desc)$"),
    limit: int | None = Query(default=None, ge=1),
    offset: int = Query(default=0, ge=0),
    allow_diff: int = Query(default=0, ge=0),
    min_duration: int | None = Query(default=None, ge=0),
    db: aiosqlite.Connection = Depends(get_db),
) -> list[JoinViolationRankOut]:
    return await repo.get_join_violation_rankings(db, group_id, start, end, order, limit, offset, allow_diff, min_duration)


@router.get("/player-rankings", response_model=list[PlayerRankOut])
async def get_player_rankings(
    world_id: str | None = Query(None),
    group_id: str | None = Query(None),
    order: str = Query(default="desc", pattern="^(asc|desc)$"),
    limit: int | None = Query(default=None, ge=1),
    offset: int = Query(default=0, ge=0),
    db: aiosqlite.Connection = Depends(get_db),
) -> list[PlayerRankOut]:
    return await repo.get_player_rankings(db, world_id, group_id, order, limit, offset)

File: server/internal/handler/analytics.go
package handler

import (
	"context"

	"github.com/njm2360/vrchat-join-manager/server/internal/gen"
	"github.com/njm2360/vrchat-join-manager/server/internal/repository"
)

func (s *Server) GetDailyActiveUsers(ctx context.Context, request gen.GetDailyActiveUsersRequestObject) (gen.GetDailyActiveUsersResponseObject, error) {
	start := timePtrToStrPtr(request.Params.Start)
	end := timePtrToStrPtr(request.Params.End)
	rows, err := s.Analytics.DailyActiveUsers(ctx, request.Params.WorldId, request.Params.GroupId, start, end)
	if err != nil {
		return nil, err
	}
	out := make(gen.GetDailyActiveUsers200JSONResponse, 0, len(rows))
	for _, r := range rows {
		out = append(out, gen.DailyActiveUsersPoint{
			Day:         r.Day,
			ActiveUsers: r.ActiveUsers,
		})
	}
	return out, nil
}

func (s *Server) GetHourlyActiveUsers(ctx context.Context, request gen.GetHourlyActiveUsersRequestObject) (gen.GetHourlyActiveUsersResponseObject, error) {
	start := timePtrToStrPtr(request.Params.Start)
	end := timePtrToStrPtr(request.Params.End)
	rows, err := s.Analytics.HourlyActiveUsers(ctx, request.Params.WorldId, request.Params.GroupId, start, end)
	if err != nil {
		return nil, err
	}
	out := make(gen.GetHourlyActiveUsers200JSONResponse, 0, len(rows))
	for _, r := range rows {
		out = append(out, gen.HourlyActiveUsersPoint{
			Hour:        r.Hour,
			ActiveUsers: r.ActiveUsers,
		})
	}
	return out, nil
}

func (s *Server) GetPlayerRankings(ctx context.Context, request gen.GetPlayerRankingsRequestObject) (gen.GetPlayerRankingsResponseObject, error) {
	order := enumStrOr(request.Params.Order, "desc")
	offset := derefInt(request.Params.Offset)
	rows, err := s.Analytics.PlayerRankings(ctx, request.Params.WorldId, request.Params.GroupId, order, request.Params.Limit, offset)
	if err != nil {
		return nil, err
	}
	out := make(gen.GetPlayerRankings200JSONResponse, 0, len(rows))
	for _, r := range rows {
		out = append(out, gen.PlayerRankOut{
			Rank:                 r.Rank,
			UserId:               r.UserID,
			DisplayName:          r.DisplayName,
			TotalDurationSeconds: r.TotalDurationSeconds,
			SessionCount:         r.SessionCount,
		})
	}
	return out, nil
}

func (s *Server) GetJoinViolationRankings(ctx context.Context, request gen.GetJoinViolationRankingsRequestObject) (gen.GetJoinViolationRankingsResponseObject, error) {
	start := timePtrToStrPtr(request.Params.Start)
	end := timePtrToStrPtr(request.Params.End)
	order := enumStrOr(request.Params.Order, "desc")
	offset := derefInt(request.Params.Offset)
	allowDiff := 0
	if request.Params.AllowDiff != nil {
		allowDiff = *request.Params.AllowDiff
	}
	rejoin := 180
	if request.Params.RejoinSeconds != nil {
		rejoin = *request.Params.RejoinSeconds
	}
	grace := 900
	if request.Params.GraceSeconds != nil {
		grace = *request.Params.GraceSeconds
	}
	rows, err := s.Analytics.JoinViolationRankings(ctx, repository.JoinViolationRankingsParams{
		GroupID:       request.Params.GroupId,
		Start:         start,
		End:           end,
		Order:         order,
		Limit:         request.Params.Limit,
		Offset:        offset,
		AllowDiff:     allowDiff,
		MinDuration:   request.Params.MinDuration,
		RejoinSeconds: rejoin,
		GraceSeconds:  grace,
	})
	if err != nil {
		return nil, err
	}
	out := make(gen.GetJoinViolationRankings200JSONResponse, 0, len(rows))
	for _, r := range rows {
		out = append(out, gen.JoinViolationRankOut{
			Rank:           r.Rank,
			UserId:         r.UserID,
			DisplayName:    r.DisplayName,
			ViolationCount: r.ViolationCount,
			TotalJoins:     r.TotalJoins,
		})
	}
	return out, nil
}
